### DISCUSSION TITLE
Node runs into a memory alarm with CQv2 and a fast publisher without publisher confirms

### DISCUSSION THREAD

**Discussion by songponlekpetch** [rabbitmq/rabbitmq-server#12885] [Other] [answered]
### Describe the bug

I plan to upgrade RabbitMQ version 3.13.7 to 4.0.4. I test with publish data without consumers for load testing. I compare the memory usage between both versions. I found version 4.0.4 that consume memory higher than version 3.13.7. You can see it from my screenshot. 

![image](https://github.com/user-attachments/assets/81c8f7f5-c165-4623-9785-14418eefce1d)

Both are use the same code and same message for data publishing.

FYI. I use classic queue version 2 with the single instance.

### Reproduction steps

1. I run this code for publish data to both instances
```go
package main

import (
	"fmt"
	"log"
	"os"
	"time"

	"github.com/sourcegraph/conc"
	"github.com/streadway/amqp"
)

func failOnError(err error, msg string) {
	if err != nil {
		log.Fatalf("%s: %s", msg, err)
	}
}

func CreatePublisherConnectionQueueV2(uri, queueName string) (*amqp.Connection, *amqp.Channel, amqp.Queue) {
	conn, err := amqp.Dial(uri)
	failOnError(err, "Failed to connect to RabbitMQ")

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channel")

	q, err := ch.QueueDeclare(
		queueName, // name
		true,      // durable
		false,     // delete when unused
		false,     // exclusive
		false,     // no-wait
		map[string]interface{}{
			"x-queue-version": 2,
		}, // arguments version 2
	)
	failOnError(err, "Failed to declare a queue")
	return conn, ch, q
}

func process(inputQueueName string, size int) {
	url := "amqp://user:password@localhost/"
	body := []byte(`{"request":{"operation":{"_collection":"project_001","_category":"source_type","_reference":"content_123_reply_456"}},"data":{"script":"ctx._source.primary_tags=primary_tags;ctx._source.secondary_tags=secondary_tags;ctx._source.excluded_tags=excluded_tags;ctx._source.image_tags=image_tags;ctx._source.searchable_text=searchable_text;if(ctx._source.status!=1&&ctx._source.status!=4){ctx._source.status=status};if(ctx._source.emotion_changed==0){ctx._source.emotion=emotion};if(ctx._source.containsKey('monitoring')){ctx._source.monitoring.content=monitoring.content;ctx._source.monitoring.user=monitoring.user;ctx._source.monitoring.category=monitoring.category}else{ctx._source.monitoring=monitoring};if(ctx._source.containsKey('match_source')){ctx._source.match_source.monitor_content=match_source.monitor_content;ctx._source.match_source.monitor_user=match_source.monitor_user;ctx._source.match_source.monitor_category=match_source.monitor_category;ctx._source.match_source.tag_match=match_source.tag_match;ctx._source.match_source.text_scan=match_source.text_scan;ctx._source.match_source.image_scan=match_source.image_scan}else{ctx._source.match_source=match_source}","params":{"modified_at":"2024-06-06T07:37:47+00:00","primary_tags":["shopping_mall","mall"],"secondary_tags":["store"],"excluded_tags":[],"image_tags":[],"searchable_text":["store"],"monitoring":{"content":0,"user":0,"category":0},"status":1,"emotion":0,"match_source":{"monitor_content":0,"monitor_user":0,"monitor_category":0,"tag_match":0,"text_scan":1,"image_scan":0}}},"fallback":{"limit":0,"count":0}}`)
	outputCon, outputCh, _ := CreatePublisherConnectionQueueV2(url, inputQueueName)
	defer func() {
		outputCh.Close()
		outputCon.Close()
	}()

	outputCloseNotify := outputCon.NotifyClose(make(chan *amqp.Error))

	go func() {
		for range outputCloseNotify {
			log.Print("Rabbit MQ connection lost")
			os.Exit(1)
		}
	}()

	var wg conc.WaitGroup
	for i := 0; i < 10; i++ {
		wg.Go(func() {
			for j := 0; j < size; j++ {
				err := outputCh.Publish(
					"",             // exchange
					inputQueueName, // routing key
					false,          // mandatory
					false,          // immediate
					amqp.Publishing{
						DeliveryMode: amqp.Persistent,
						ContentType:  "text/plain",
						Body:         body,
					})

				failOnError(err, "Failed to publish a message")
			}
		})
	}
	fmt.Println("Waiting for all goroutines to finish")
	wg.Wait()
	fmt.Println("All for 10 seconds before closing the connection")
	// sleep for 10 secs
	time.Sleep(10 * time.Second)

}

func main() {
	for i := 0; i < 2; i++ {
		go process(fmt.Sprintf("test_%d", i), 100000000)
	}
	forever := make(chan bool)
	<-forever
}

```
2. It will create 2 classic queue version 2 as test_0 and test_1 

### Expected behavior

Version 4.0.4 should not consume higher memory than version 3.13.7

### Additional context

I test both versions on instances that run on Ubuntu version 24.04 LTS in AWS EC2 instance type m6a.large.

**Comment by mkuratczyk** [comment]
I'm not sure why the difference is so significant, but the main problem is that this code doesn't use publisher confirms and therefore just overloads RabbitMQ. 

unrelated to the question, but you should switch to https://github.com/rabbitmq/amqp091-go, which is a continuation of github.com/streadway/amqp

**Comment by mkuratczyk** [comment]
I will investigate it to see why this happens, but as mentioned above - generally speaking, your code should be using confirms and this wouldn't have happened.

For the record, repro using https://perftest.rabbitmq.com/:
`perf-test -ad false -f persistent -qp cq-%d -qpf 1 -qpt 2 -x 2 -y 0 -qa x-queue-version=2`

**Reply by mkuratczyk** [reply]
Update from my investigation so far:
1. While perf-test works for repro, you Go code works "better" (it's even more aggressive so it's easier to trigger the problem)
2. The commit that changes this behaviour is the removal of mirrored queues: https://github.com/rabbitmq/rabbitmq-server/commit/ae31b47d0b691e42795ca65983b39205c9d0670e

Changes to the flow control (ultimately that seems to be the problem - before this change, flow control prevented queue process overload, afterwards it doesn't) is an unexpected side effect of the removal of queue mirroring.

**Reply by michaelklishin** [reply]
Removing flow control was not the goal but since CMQ had it in two different places IIRC (it was easier to overwhelm than a non-replicated queue), I guess both ended up being casualties of the CMQ removal, even though one could stay (the one that was not between replicas).

**Reply by gomoripeti** [reply]
we noticed a few weeks ago looking at 4.0 behaviour the removal of credit_flow between channel and CQ process in the above mentioned commit. We did not understand the connection between flow control and removal of mirroring, but thought it is replaced by some newer "credit API version N+1" mechanism. So it was not a deliberate change (to keep `delivery_flow = undefined` https://github.com/rabbitmq/rabbitmq-server/commit/ae31b47d0b691e42795ca65983b39205c9d0670e#diff-ab2459b0afd286f14f2f08978ddbc9c55ee927f933eb3d687a4f0bd4a8c572c6R162) ?

**Reply by michaelklishin** [reply]
The credit API first and foremost applies to AMQP 1.0. Which is arguably the core of RabbitMQ more than AMQP 0-9-1 these days due to Message Containers using the AMQP 1.0 serialiation format and AMQP 1.0 being a built-in protocol now.

What I'm saying is that flow control for AMQP 0-9-1 channels likely wasn't at all in mind by the folks who work on the AMQP 1.0 credit flow and settlement areas.

**Reply by michaelklishin** [reply]
@gomoripeti you are welcome to re-introduce the channel-queue flow control. At the very least we will consider it after load testing. If it did make a meaningful difference, it can be brought back.

**Reply by gomoripeti** [reply]
thank you!

**Reply by lhoguin** [reply]
This one relates to `mirroring_flow_control` and was I believe completely ignored by plain CQs (in publish1).

**Reply by lhoguin** [reply]
Credit flow with the channel is still around: https://github.com/rabbitmq/rabbitmq-server/commit/ae31b47d0b691e42795ca65983b39205c9d0670e#diff-0829cdc953ddf9e3363c42e6e3503a95a6d28fd896fe0f99d6eba8b0b204e661R390

**Reply by gomoripeti** [reply]
the problem is that Flow is always `noflow` https://github.com/rabbitmq/rabbitmq-server/commit/ae31b47d0b691e42795ca65983b39205c9d0670e#diff-0829cdc953ddf9e3363c42e6e3503a95a6d28fd896fe0f99d6eba8b0b204e661R379 because it is always undefined in Options
(I was also surprised that all the code for plain classic queue credit_flow is still around but unused)

**Reply by lhoguin** [reply]
I'm seeing it now. I think you got the culprit, should be easy enough to confirm (if not confirmed already).

**Reply by gomoripeti** [reply]
with the perf-test command repro Michal provided it is the internal message queue of the gen_server2 of the queue process that is growing unbounded.
Did not have time to test this draft patch yet (https://github.com/cloudamqp/rabbitmq-server/commit/e80ec7d38ed7eb825a8ac60fd21c126388c4750f) but I intend to get back to this task as soon as I can.

**Reply by mkuratczyk** [reply]
Unfortunately this patch doesn't fix it for me. Also, I'll just re-iterate that I'm using the code from this issue as a test as it is more "reliable" (perf-test sometimes triggers it and sometimes doesn't for me)

**Reply by gomoripeti** [reply]
thanks Michal for testing. Yes I realise there were more things removed in `rabbit_classic_queue:deliver` and probably in the backing queue modules as well. (But that might mean some changes in the backing_queue callbacks where the Flow argument was removed might have to be reintroduced, which is a braking change and cannot be done just in a patch release)

**Reply by gomoripeti** [reply]
Fortunately it was not necessary to touch the backing queue behaviour.
FTR I opened the PR https://github.com/rabbitmq/rabbitmq-server/pull/12906 with the changes.

**Comment by michaelklishin** [comment]
@songponlekpetch you haven't provided any evidence of a bug.

There is a [dedicated documentation guide](https://www.rabbitmq.com/docs/memory-use) and that explains how to collect data on what uses the memory.

@mkuratczyk's hypothesis seem right. A publisher does drastically less work than RabbitMQ, so without publisher confirms it will eventually drive the node above the threshold after which an alarm is triggered and publishers are blocked to protect the node.

If you need fire-and-forget publishing with maximum throughput and a mostly flat memory footprint, use streams or superstreams with a dedicated RabbitMQ Stream Protocol client https://github.com/rabbitmq/rabbitmq-stream-go-client.

Quorum queues, as their docs [explain](https://www.rabbitmq.com/docs/quorum-queues#resource-use), have a certain footprint that is usually stable within a range, so it looks like a zig-zag pattern. Even with [tuning](https://www.rabbitmq.com/docs/quorum-queues#performance-tuning) this won't change fundamentally.

Classic queues, even v2, are less predictable than the other two types and using publisher confirms will be particularly important.

**Comment by michaelklishin** [comment]
> I found version 4.0.4 that consume memory higher than version 3.13.7.
> You can see it from my screenshot.

I am sorry but this proves absolutely nothing about the memory footprint of CQs in both versions.
You have a fire-and-forget, go as-fast-as-the-resources-allow publisher that is guaranteed to run into an alarm one way or another.

A single run of such publisher proves absolutely nothing. [Long-running, more realistic workloads](https://www.rabbitmq.com/blog/tags/performance) with metrics collected over the entire duration of the run do.

**Comment by michaelklishin** [comment, marked as answer]
@songponlekpetch thanks to @gomoripeti, an unintentionally removed portion of classic queue flow control was restored in https://github.com/rabbitmq/rabbitmq-server/pull/12906, https://github.com/rabbitmq/rabbitmq-server/pull/12907.

You can get a new development version to try from https://github.com/rabbitmq/server-packages/releases/tag/alphas.1733843440418. A `4.0.x` version will be available in the same repo when we backport the changes. The plan right now is to ship this change in `4.0.5`.

**Reply by michaelklishin** [reply]
[`4.0.5-alpha.ebfeab82`](https://github.com/rabbitmq/server-packages/releases/tag/alphas.1733860582115) is a `4.0.5` preview that includes #12906 #12907.

**Reply by songponlekpetch** [reply]
Thank you so much for your support.

**Reply by michaelklishin** [reply]
@songponlekpetch does the above version match (more or less) the desired behavior for you?

**Reply by songponlekpetch** [reply]
Yes, it does. After I stress test, it doesn't consume high memory like before. Thanks @michaelklishin 
![image](https://github.com/user-attachments/assets/f526440c-74cb-4736-bbb4-3970293d5852)
